Improve length handling in string preprocessing

allredj opened the pull request:
Partially replaces #984.
Partially addresses diffblue/test-gen#256.

Depends on #993 (so only the last commit of this PR needs reviewing).

allredj commented:
At the moment, this introduces a regression in two of the test-gen regression tests. So the status should be DON'T MERGE.

smowton commented on src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1013,42 +1013,81 @@ string_exprt java_string_library_preprocesst::
+  const struct_typet &jlo_struct=to_struct_type(jlo_symbol.type);
+  struct_exprt jlo_init(jlo_struct);
+  jlo_init.copy_to_operands(constant_exprt(
+    "java::java.lang.String", jlo_struct.components()[0].type()));
```
Recommend using https://github.com/diffblue/cbmc/blob/master/src/java_bytecode/java_object_factory.cpp#L264 to do this, to avoid duplicating implementations of java.lang.Object init and automatically keep up to date if its format changes.

allredj replied:
@smowton I see, but we don't have an instance of `java_object_factory` at this place. How do you think we should we do this?

allredj replied:
Ping @smowton .

NlightNFotis commented on src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1772,6 +1864,82 @@ codet java_string_library_preprocesst::
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
+  (void) loc; // Avoid compiler warning for unused parameter
```
Would it not be better to add some error handling to this method, and use the `loc` variable in the case of an error? The code below makes the assumption that all operations are fail-safe. Is it so? Nothing can throw an exception?

allredj replied:
This code has now changed, and the `loc` variable is used. I don't think there are any exceptions that we can handle at this stage.

NlightNFotis commented on src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -1772,6 +1864,82 @@ codet java_string_library_preprocesst::
+  const code_typet &type,
+  const source_locationt &loc,
+  symbol_tablet &symbol_table)
+{
```
I am also noticing the same thing here: Perhaps it's my unfamiliarity with the codebase, but this is a linear progression of operations that to my understanding assume fail-safe -ness. Is it so?

allredj replied:
I'm not sure I understand what you mean. What kind of failure are you talking about?

allredj commented on src/java_bytecode/java_string_library_preprocess.cpp:
```diff
@@ -2056,18 +2244,27 @@ void java_string_library_preprocesst::initialize_conversion_table()
+      "Ljava/lang/StringBuilder;"]=
+      ID_cprover_string_concat_func;
+  cprover_equivalent_to_java_assign_and_return_function
+    ["java::java.lang.StringBuilder.append:(Ljava/lang/CharSequence;)"
```
The first one is for signature:
```java
public StringBuilder append(CharSequence s)
```
and the second one is for signature:
```java
public StringBuilder append(CharSequence s,
                            int start,
                            int end)
```

allredj commented:
I have squashed a new version of `java_string_library_preprocesst::make_string_length_code` that creates a `string_exprt`. This is necessary for the  string solver to take it into account during string refinement. The rest of the code hasn't changed, but I have corrected/added some regression tests.

allredj commented:
@NlightNFotis any further comments?

NlightNFotis commented:
@allredj Not really, no. Looks good to me now.
